### DISCUSSION TITLE
Marking up code blocks for unsupported languages

### DISCUSSION THREAD

**Discussion by hamishwillee** [mdn/mdn-community#170] [Content] [answered]
What should our rules be for marking up code blocks when the languages are not one of the supported set (as listed in https://github.com/mdn/yari/blob/main/build/syntax-highlight.js#L15).

Options:
1. Add support for all language(s).
2. Use `plain` to indicate that the type is explicitly set as nothing/unsupported
3. Use a substitute language where you get somewhat useful highlighting - e.g. `bash` as a substitute for `powershell`, and use `plain` otherwise.
4. Markup in the actual language of the text, whether or not it renders. Use `plain` if the "code" doesn't have a language, or at least one supported by prism.

The ideal would be to do "1" but it isn't clear what the cost of adding a new language is. If it is a per-page cost then you'd reasonably just add all the languages, whereas if the cost applies to every page, you'd have to consider pages separately. Haven't got an answer on how this works - trying in https://github.com/mdn/yari/issues/6763

Either way, currently I use "3" because really, the syntax for all the shells is much the same (so you get some benefit using bash for command, powershell, etc etc.). But I'm leaning towards suggesting we use "4"). 

Why? My assumption is that this is the best future proofing and indication of intent. It can't be wrong, and it should all work nicely if we were to add linters for different language types later on.

[Note, posted this because of https://github.com/mdn/content/pull/19259

**Comment by queengooborg** [comment]
Considering that the file for importing and loading Prism is in the `build/` folder, and skimming the source code shows that syntax highlighting is performed during the build step, I'm voting for option 1 as there would be no (or _very_ minimal) cost on the reader's end.

**Reply by queengooborg** [reply]
I've opened https://github.com/mdn/yari/pull/6831 to add many more languages (and add some comments to list out the languages loaded by default, so the function can be used as a reference of available languages)!

**Comment by wbamberg** [comment]
Whatever we do, please let's document the identifiers that authors should be allowed to use, and update the list when there is a need for more.

We currently do this here: https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Howto/Markdown_in_MDN#example_code_blocks. I'm not really in favour of people being able to add whatever identifiers they feel like. There isn't a cost to readers in this but there is a cost to maintainers.

**Reply by hamishwillee** [reply]
Ah, so you think we should use "2" as opposed to "4" - i.e. have a discreet subset and mark the things we don't support with `plain`. I think that's OK too. Possibly also have a note explaining how/where you might request updates to yari.

**Reply by hamishwillee** [reply]
@wbamberg If we're not allowing everything then we need to
- agree a policy for what we will allow/want, 
- accept and document an initial set that matches that. 
- add the policy and process for getting new ones added to the docs.

Note, github uses a very large set: https://github.com/github/linguist/blob/master/lib/linguist/languages.yml
We should use a subset of those.

Not sure what the policy should be. Any ideas? I would argue around
- dominant language of a supported reference set (e.g. if we document typescript or wasm we would have it)
- makes code markup significantly easier to read than `plain` for some significant set of docs
- discuss on a case by case basis by posting justification as a PR to yari.

The list proposed in @queengooborg initial PR is below. I would argue for wasm and jsx based on first point and powershell because this gets added by users enough to be irritating. I'm not sure about yaml - it is pretty well served by "plain". `pug` pages would definitely benefit from this, but we're only talking 3 or 4 of them - is that wortht he cost? 

```
  "powershell",
   "jsx",
    "typescript",
    "diff",
    "ini",
    "ignore",
    "less",
    "regex",
    "scss",
    "sql",
    "uri",
    "wasm",
    "yaml",
    "pug",
```

**Comment by Josh-Cena** [comment]
So according to @queengooborg adding a language is virtually zero-cost for the user. In this case I don't see any reason why we can't mark up all code blocks with the right language. In my editor, the code will be displayed with the corresponding syntax highlighting; for static analysis tools (not just limited to JS linters), they will be able to grep all code blocks to be linted; for users, they enjoy richer syntax highlighting, or failing that, at least with no adverse effects.

I'm going towards "4". We should do the following:

- Be as correct as possible (i.e. not arbitrarily tell people "no, we don't use `typescript` on MDN, please just mark it as `js`")
- Whenever there's a need for a new language, go add it in Yari
- _However_, before the language becomes available in Yari or before it's on its way to becoming available, do not allow people to tag code blocks with languages that are never to be enabled (usually because Prism doesn't support it—e.g. `vue`/`svelte`).

I don't have strong opinions on telling people what language should be allowed to use. If we have a diverse language set to start with, that definitely works. What I don't want is for people to work around the platform limitations and mark up powershell code blocks as bash.

**Reply by hamishwillee** [reply]
@wbamberg Perhaps you could explain further why you think supporting all languages increases our maintenance cost? I have no strong opinion either way.

**Reply by hamishwillee** [reply]
I will raise one item - "if you build it they will come". If we add everything then people may assume that means we want them to use it. But as per https://github.com/mdn/yari/pull/6831/files#r944176341 we really don't want people documenting the IDL, for example.

**Reply by Josh-Cena** [reply]
"Don't write WebIDL" should then be part of the style guide explicitly, instead of doing so via a technical constraint. It's only one case—I'm pretty sure all other code serve a purpose in the docs. For example, we wouldn't have dreamed of supporting LaTeX before we had MathML guides that actually show TeX source.

Since we have a thorough review process now, maintaining a blocklist of languages also seems easier for the writers than an allowlist, where the allowlist can get really long (provided that "try to use the right language tag" is a goal).

**Reply by wbamberg** [reply]
As an author I want to know which language identifiers I can us to mark up code blocks. At the moment AIUI this list is at https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Howto/Markdown_in_MDN#example_code_blocks . Note that this list doesn't just list the languages but also the identifier (i.e. "js" not "javascript"). 

I'm not totally sure what the proposal is here. Is it "expand the set to all languages/identifiers that Prism supports"? If so, I think we need to provide that set, or a stable link to it. And if we stop using Prism, does the set change?

But the OP also says "Add support for all language(s)". Really all languages? And what is the identifier for each language? I guess in https://github.com/github/linguist/blob/master/lib/linguist/languages.yml they have aliases, is that part of the proposal? Does that mean people can mark up JS code blocks with "js" or "javascript"? It seems like this will be confusing to people reading and editing the source (this is like how in Markdown we'd like to ensure people use [ATX headings](https://spec.commonmark.org/0.30/#atx-headings) rather than [setext headings](https://spec.commonmark.org/0.30/#setext-headings), even if our tooling could handle both).

One of the diseases of the MDN platform is that there are many ways to do the same thing, so everyone does it their way, and the source becomes more complicated and harder to write. I would like us to keep that in mind, and work towards making this less of an issue.

So I'm not against adding new languages that we need. But I think if we do, we have a responsibility to document them.

**Reply by hamishwillee** [reply]
I think we're close to some kind of consensus: no one "really" wants all languages - we want a clear path for the useful ones to be added and people to be able to know the set available and how to use them. 

So combination of @Josh-Cena post above + requirement to document (@wbamberg):

- We _allow_ any language supported by prism except for the blocklist (blockslist is currently just WebIDL).
- We don't add "all languages" by default. Someone has to "want" them and be willing to document them.
- Users can request a new language via a Yari PR, noting:
  - the language and identifier used on MDN.
    - Identifier should match Github - we're trying to be close to GFM.
    - If there are two possible identifiers, the most common contraction should be documented (and enforced) on MDN.
    - The identifier should be added to the alias list even if prism already supports it "natively". This allows the list of identifiers to be inferred from the alias list.
  - a link to the language definition (i.e. for ini it is not obvious what the language might be)
  - a non-spurious reason<sup>++</sup>
  - Link to any places known to be affected in MDN.
- The _PR author_ is responsible for also adding alias, language and link to language definition to [the list](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Howto/Markdown_in_MDN#example_code_blocks).
- Unsupported languages do not get submitted - they must be set as plain or wait for corresponding PR to be accepted.

<sup>++</sup>The "non-spurious" reason should be pretty lax. I just want PR reviewers to have a right to reject PRs where no thought has gone into adding the language. Should we have "Fortran"? Maybe, but not because someone decides it might be fun.

**Reply by Josh-Cena** [reply]
Yep, I agree with @hamishwillee's proposal. We "allow any language by principle" (except those explicitly in blocklist), but the language must be supported in Yari first to be used in content. That also depends on how quickly PRs to Yari are merged, though—I didn't have great experience with the review process on Yari so far 🙃 I hope some day we can maintain the Prism language list on content side.

**Reply by wbamberg** [reply]
This sounds great to me, Hamish, thank you for thinking it out. I would call it "identifier" rather than alias, because to me "alias" implies an alternate name, and I think we want to have only one name. Otherwise +1

**Reply by Josh-Cena** [reply]
Also, FWIW, `js` and `javascript` are already both supported—as soon as we load one language, all aliases of it defined by Prism are available out-of-the-box.

**Reply by hamishwillee** [reply]
@wbamberg @Josh-Cena 
- Awesome. I modified the text above to say "identifier" rather than alias.
- > Also, FWIW, js and javascript are already both supported—as soon as we load one language, all aliases of it defined by Prism are available out-of-the-box.

  Good point.  But I think we should settle on documenting just one identifier because it may make things easier if we want to lint the sections, and it certainly makes things easier for reviewers to have just one valid case. Not to mention possible future migration to something other than prism. Lastly, I can't see any real benefit in stating you use one of these identifiers to indicate a language.
  
  Above I have updated this to add sub rules
  - Identifier should match Github - we're trying to be close to GFM.
  - If there are two possible identifiers, the most common contraction should be documented (and enforced) on MDN.
  - The identifier should be added to the alias list even if prism already supports it "natively". This allows the list of identifiers to be inferred from the alias list.

I'll try apply this rule to https://github.com/mdn/yari/pull/6831/files and see how it turns out.

**Reply by teoli2003** [reply]
Looks good to me. @hamishwillee: the answer is a bit lost in the discussion right now. Can you copy it into a new _Suggested answer_ below? Then, if the consensus stays, mark it as the official answer. That way, we won't need to go through the whole discussion to refer to it.

**Reply by hamishwillee** [reply]
Done!

**Comment by hamishwillee** [comment, marked as answer]
The consensus appears to be:

- We _allow_ any language supported by prism except for the blocklist (blockslist is currently just WebIDL).
- We don't add "all languages" by default. Someone has to "want" them and be willing to document them.
- Users can request a new language via a [Yari PR](https://github.com/mdn/yari), noting:
  - the language and identifier used on MDN.
    - Identifier should match Github - we're trying to be close to GFM.
    - If there are two possible identifiers, the most common contraction should be documented (and enforced) on MDN.
    - The identifier should be added to the alias list even if prism already supports it "natively". This allows the list of identifiers to be inferred from the alias list.
  - a link to the language definition (i.e. for ini it is not obvious what the language might be)
  - a non-spurious reason<sup>++</sup>
  - Link to any places known to be affected in MDN.
- The _PR author_ is responsible for also adding alias, language and link to language definition to [the list](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Howto/Markdown_in_MDN#example_code_blocks).
- Unsupported languages do not get submitted - they must be set as plain or wait for corresponding PR to be accepted.

<sup>++</sup>The "non-spurious" reason should be pretty lax. I just want PR reviewers to have a right to reject PRs where no thought has gone into adding the language. Should we have "Fortran"? Maybe, but not because someone decides it might be fun.

**Comment by Aldo-Solares** [comment]
A .env would be nice
